api/audit/tasks.py

```diff
@@ -150,16 +161,114 @@ def create_segment_priorities_changed_audit_log(
         environment=environment,
-        project_id=environment.project_id,
+        log=f"Segment overrides re-ordered for feature '{feature.name}'.",
         author_id=user_id,
-        related_object_id=feature.id,
-        related_object_type=RelatedObjectType.FEATURE.name,
         master_api_key_id=master_api_key_id,
+        related_object_id=feature.pk,
+        related_object_type=RelatedObjectType.FEATURE.name,
+    )
+
+
+@register_task_handler()
+def create_audit_log_user_logged_in(
+    user_id: int,
+    ip_address: str | None = None,
+):
+    if not (user := get_user_model().objects.filter(id=user_id).first()):
+        logger.warning(
+            f"User with id {user_id} not found. Audit log for user logged in not created."
+        )
+        return
+
+    user = typing.cast(_AbstractBaseAuditableModel, user)
+    log_message = (
+        f"{RelatedObjectType.USER.value} logged in: {user.get_audit_log_identity()}"
     )
+
+    audit_logs = [
+        AuditLog(
+            organisation=organisation,
+            related_object_id=user.pk,
+            related_object_type=RelatedObjectType.USER.name,
+            log=log_message,
+            author=user,
+            ip_address=ip_address,
+            is_system_event=True,
+        )
+        for organisation in user._get_organisations() or []
+    ]
+    AuditLog.objects.bulk_create(audit_logs)
+
+
+@register_task_handler()
+def create_audit_log_user_logged_out(
+    user_id: int,
+    ip_address: str | None = None,
+):
+    if not (user := get_user_model().objects.filter(id=user_id).first()):
+        logger.warning(
+            f"User with id {user_id} not found. Audit log for user logged out not created."
+        )
+        return
+
+    user = typing.cast(_AbstractBaseAuditableModel, user)
+    log_message = (
+        f"{RelatedObjectType.USER.value} logged out: {user.get_audit_log_identity()}"
+    )
+
+    audit_logs = [
+        AuditLog(
+            organisation=organisation,
+            related_object_id=user.pk,
+            related_object_type=RelatedObjectType.USER.name,
+            log=log_message,
+            author=user,
+            ip_address=ip_address,
+            is_system_event=True,
+        )
+        for organisation in user._get_organisations() or []
+    ]
+    AuditLog.objects.bulk_create(audit_logs)
+
+
+@register_task_handler()
+def create_audit_log_user_login_failed(
+    credentials: dict,
+    ip_address: str | None = None,
+    codes: list[str] | str | None = None,
+):
+    if not (username := credentials.get("username")):
+        return
+    UserModel = get_user_model()
+    try:
+        # ModelBackend looks up user this way, so we will do the same
+        user = get_user_model()._default_manager.get_by_natural_key(username)
+    except UserModel.DoesNotExist:
+        return
+    if not isinstance(user, _AbstractBaseAuditableModel):
+        return
+
+    codes = [codes] if type(codes) is str else codes
```

Comment:
As far as I can see it's a bug in DRF - for some exceptions `get_codes()` returns `str` and others it's `list[str]` 😞. I was 50/50 where to put the workaround, I'll push it back to the signal handler rather than the task.